backend/src/main/java/ai/giskard/web/rest/controllers/ProjectController.java

```diff
@@ -132,4 +142,29 @@ public ProjectDTO invite(@PathVariable("id") Long id, @PathVariable("userId") Lo
         return giskardMapper.projectToProjectDTO(project);
     }
 
+    @PostMapping(value = "/project/import/{timestampDirectory}/{projectKey}/prepare", consumes = "multipart/form-data")
+    public PrepareImportProjectDTO prepareImport(@RequestParam("file") MultipartFile zipFile, @PathVariable("timestampDirectory") @NotNull String timestampDirectory, @PathVariable("projectKey") String projectKey, @AuthenticationPrincipal final UserDetails userDetails) throws IOException {
+        Path pathToTimestampDirectory;
+        permissionEvaluator.canWrite();
+        pathToTimestampDirectory = projectService.unzip(timestampDirectory, zipFile);
+        try {
+            return projectService.prepareImport(pathToTimestampDirectory, projectKey);
+        } catch (IOException e ) {
+            FileSystemUtils.deleteRecursively(pathToTimestampDirectory);
+            throw new GiskardRuntimeException("Error while preparing your project import", e);
+        }
+    }
+
+    @PostMapping(value = "/project/import/{timestampDirectory}/{projectKey}")
+    @Transactional
+    public ProjectDTO importProject(@RequestBody Map<String, String> mappedUsers, @PathVariable("projectKey") @NotNull String projectKey, @PathVariable("timestampDirectory") @NotNull String timestampDirectory, @AuthenticationPrincipal final UserDetails userDetails) throws IOException {
+        permissionEvaluator.canWrite();
+        try {
+            return projectService.importProject(mappedUsers, timestampDirectory, projectKey, userDetails.getUsername());
+        } catch (IOException e){
+            throw new GiskardRuntimeException("Error while importing the project", e);
+        }finally {
+            FileSystemUtils.deleteRecursively(locationService.resolvedTmpPath());
```

Comment:
the temporary directory cleaning should be part of the service, not the controller. Ideally controllers should be responsible for triggering actions and formatting responses according to the chosen format (DTO serialized to JSON in our case)

Reply:
my change is better now or still not what you expect ?

Reply:
in JAVA it's not typical to use the `__` method prefix. Private methods should just have a `private` modifier. Plus I guess you could have a feeling that the original `importProject` method was too large to be wrapped with additional try-catch-finally. A better solution instead of wrapping one `importProject` with another would be to only keep a single `importProject` with `try` block but extract each entity creation into a separate submethod like `importModels`, `importDatasets`. By doing so you'll also keep the main `importProject` method on a higher level of abstraction than the methods it'll be calling which is a good practice.